model/installation_request.go

```diff
@@ -90,6 +94,32 @@ func (request *CreateInstallationRequest) Validate() error {
 	return checkSpaces(request)
 }
 
+func isValidDNS(dns string) error {
+	subdomain := strings.SplitN(dns, ".", 2)[0]
+	if len(subdomain) >= 64 || len(subdomain) < 3 {
+		return errors.Errorf("DNS names must be between 3 and 64 characters, but name was %d long. DNS=%s", len(dns), dns)
+	}
+	// check that domain matches regex for valid names
+	if found := hostnamePattern.FindString(dns); found != dns {
+		return errors.Errorf("DNS name provided (%s) failed hostname pattern check", dns)
+	}
+	// check that domain does not resolve. Use a custom pure-Go resolver
+	// to get the same behavior on VPN, in testing, and in production
+	r := &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			return (&net.Dialer{
+				Timeout: time.Second * time.Duration(10000),
+			}).DialContext(ctx, "udp", "1.1.1.1:53")
+		},
+	}
+	_, err := r.LookupHost(context.Background(), dns)
+	if err == nil {
```

Comment:
So, I guess it is possible that the provided domain exists, but this lookup encounters some other type of error which results in the domain-taken check passing.

This is probably pretty unlikely to happen, but wanted to note it. Can we check the error type perhaps to also fail when it was a network-related error?

Reply:
Let me see what I can do

Reply:
@gabrieljackson, looking at the Go source code, it looks like the only error type that `net.LookupHost` produces is `net.DNSError`.

Here is an example of an error from looking up a nonexistent domain:
```
gore> net.LookupHost("foos.party")
[]string{}
&net.DNSError{
  Err:         "no such host",
  Name:        "foos.party",
  Server:      "10.247.0.2:53",
  IsTimeout:   false,
  IsTemporary: false,
  IsNotFound:  true,
}
```
(off topic, `gore` is a pretty neat Go REPL if you hadn't seen that before)

Here's an example of an error caused by network connectivity (`foo.party` is a registered domain):
```
gore> net.LookupHost("foo.party")
[]string{}
&net.DNSError{
  Err:         "read udp [::1]:60216->[::1]:53: read: connection refused",
  Name:        "foo.party",
  Server:      "[::1]:53",
  IsTimeout:   false,
  IsTemporary: true,
  IsNotFound:  false,
}
```

So, we can't act on the type of error, but I could provide a different error if `IsNotFound` is `false` and either `IsTimeout` or `IsTemporary` is `true`. How does that sound?

Reply:
That sounds great!

Reply:
![Screenshot from 2020-09-22 13-11-12](https://user-images.githubusercontent.com/3743903/93920868-658a0880-fcd5-11ea-8c92-b0ffe185c952.png)
it'd be a real PITA to write a test for this case, so here's a screenshot of me running the existing tests but with my network disconnected :stuck_out_tongue_closed_eyes: 

If y'all think this warrants a proper test I can figure out how to simulate a network disconnect to add a test case but it sounds like a fair amount of work with little gain.

Reply:
Haha, this is more than fine for now. Good work.